File: python/code_generator.py
'''
    Instruction Reference From:
    https://en.wikipedia.org/wiki/X86_instruction_listings
'''
class code_generator:
    def __init__(self, name):
        self.stack = 4
        print("Calling Code Generator on file " + name)
        self.output_file = open(name + '.s', 'w')
        self.label = -1

    def write_code_to_file(self, str):
        self.output_file.write(str+"\n")

    def instructions(self, operation):
        if operation == '+':
            return 'add'
        if operation == '-':
            return 'sub'
        if operation == '++':
            return 'inc'
        if operation == '--':
            return 'dec'
        if operation == '*':
            return 'imul'
        if operation == '/':
            return 'idiv'
        if operation == '>':
            return 'setg'
        if operation == '<':
            return 'setl'
        if operation == '<=':
            return 'setle'
        if operation == '>=':
            return 'setge'
        if operation == '!=':
            return 'setne'
        if operation == '==':
            return 'sete'
        pass

    def form_label(self):
        self.label += 1
        return 'L'+str(self.label)

    def write_label(self, label):
        print("label ::::::::::::",label)
        self.write_code_to_file(label + ":")

    '''
        Referred This for function Prologue:
        https://en.wikipedia.org/wiki/Function_prologue

    '''

    def write_function_begin(self, name, params, label):
        self.write_code_to_file(".global " + name)
        self.write_code_to_file(name +":")
        self.write_code_to_file("\t" + "push ebp")
        self.write_code_to_file("\t" + "mov ebp, esp")
        self.write_code_to_file("\t" + "sub esp, " + str(params * 4))
        self.write_code_to_file("\t" + "jmp " + label)
    
    def write_function_end(self):
        self.write_code_to_file("\t" + "mov esp, ebp")
        self.write_code_to_file("\t" + "pop ebp")
        self.write_code_to_file("\t" + "ret")

File: go_lang/compiler/code_generator.go
package compiler

import (
	"io/ioutil"
	"log"
	"strconv"
)

type CodeGenerator struct {
	stack      int
	outputFile string
	label      int
}

func (c CodeGenerator) initialize(name string) {
	c.stack = 4
	c.outputFile = name
	c.label = -1
}

func (c CodeGenerator) write_code_to_file(str string) {
	err := ioutil.WriteFile(c.outputFile, []byte(str), 0666)
	if err != nil {
		log.Fatal(err)
	}
}

func (c CodeGenerator) instructions(operation string) string {
	switch operation {
	case "+":
		return "add"
	case "-":
		return "sub"
	case "*":
		return "imul"
	case "/":
		return "idiv"
	case ">":
		return "setg"
	case "<":
		return "setl"
	case ">=":
		return "setge"
	case "<=":
		return "setle"
	case "!=":
		return "setne"
	case "==":
		return "sete"
	case "++":
		return "inc"
	case "--":
		return "dec"
	}
	return ""
}
func (c CodeGenerator) formLabel() string {
	c.label++
	return "L" + strconv.Itoa(c.label)
}

func (c CodeGenerator) writeLabel(label string) {
	c.write_code_to_file(label + ":\n")
}

func (c CodeGenerator) functionBegin(name string, params int, label string) {
	c.write_code_to_file(".global " + name + "\n")
	c.write_code_to_file(name + ":\n")
	c.write_code_to_file("\t" + "push ebp\n")
	c.write_code_to_file("\t" + "mov ebp, esp\n")
	c.write_code_to_file("\t" + "sub esp, " + strconv.Itoa(params*4) + "\n")
	c.write_code_to_file("\t" + "jmp" + label + "\n")
}

func (c CodeGenerator) functionEnd() {
	c.write_code_to_file("\t" + "mov esp, ebp\n")
	c.write_code_to_file("\t" + "pop ebp\n")
	c.write_code_to_file("\t" + "ret\n")
}
